Random Groups (09/24 class):
--

1. "Killian", "Sean", "Caitlin"
2. "Chase", "Luis", "Olsi"
3. "James S", "Flavio", "Lauren"
4. "Amy", "James H", "Michael"
5. "Max", "Jennifer", "Jacob"
6. "Sam", "Eli", "Cameron"


https://student.desmos.com/?prepopulateCode=x48wcx

---
# Section 1.3: Triangular Systems
---

Let $L \in \mathbb{R}^{n \times n}$ be a **lower-triangular matrix** and $U \in \mathbb{R}^{n \times n}$ be an **upper-triangular matrix**. Then,

$$
L = 
\begin{bmatrix}
l_{11} \\
\vdots & \ddots \\
l_{n1} & \cdots & l_{nn}
\end{bmatrix}
\qquad \text{and} \qquad
U = 
\begin{bmatrix}
u_{11} & \cdots & u_{1n} \\
& \ddots & \vdots \\
& & u_{nn}
\end{bmatrix}.
$$

> ### Theorem:
> Let $G$ be an $n \times n$ triangular matrix. Then
> 
> $$ 
\text{$G$ is nonsingular} \quad \Longleftrightarrow \quad g_{ii} \neq 0, \quad i = 1,\ldots,n.
$$

---
### Exercise: 

Prove the above theorem.

### Proof:

Since $G$ is triangular, the determinant of $G$ is the product of its diagonal entries; that is, $\det(G) = g_{11} g_{22} \cdots g_{nn}$. Thus, $G$ is nonsingular if and only if $g_{11} g_{22} \cdots g_{nn} \ne 0$, which holds if and only if $g_{ii} \ne 0$ for $i = 1,\ldots,n$. $\blacksquare$

---

### Exercise:
Solve the following lower-triangular system $Lx = b$ and use `Julia` to check your answer.

$$
\begin{bmatrix}
2 & 0 & 0 & 0 \\
-1 & 2 & 0 & 0 \\
3 & 1 & -1 & 0 \\
4 & 1 & -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
2 \\ 3 \\ 2 \\ 9
\end{bmatrix}.
$$

In [23]:
L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

x = L\b

4-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0

---

In general, the solution of a lower-triangular system $Lx = b$ is given by

$$
x_i = \left( b_i - \sum_{j=1}^{i-1} l_{ij}x_j \right) \bigg/ l_{ii}, \quad i = 1, \ldots, n.
$$

---

### Exercise:

Write a `Julia` function, `x = row_fs(L, b)`, that returns the solution $x$ of a lower-triangular system $Lx = b$ using _forward substitution_. Make sure to include a check that the matrix $L$ is nonsingular. Test your code for accuracy and efficiency.

In [24]:
using LinearAlgebra

function row_fs(L, b)
    
    if !istril(L)
        error("L is not lower triangular")
    end
    
    n = size(L,1)
    
    x = copy(b) # OR: x = zeros(n)
    
    for i = 1:n
        # x[i] = b[i] # if x initialized as zeros(n)
        for j = 1:i-1
            x[i] -= L[i,j]*x[j]
        end
        if L[i,i] == 0.0
            error("L is singular")
        end
        x[i] /= L[i,i]
    end
    
    return x
end

L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

x = row_fs(L, b)

4-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0

In [25]:
typeof(L)

Array{Float64,2}

In [26]:
istril(L)

true

In [27]:
methods(istril)

# 14 methods for generic function "istril":
[1] istril(A::BitArray{2}) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/bitarray.jl:165
[2] istril(M::SymTridiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/tridiag.jl:274
[3] istril(M::Tridiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/tridiag.jl:575
[4] istril(A::LowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:281
[5] istril(A::UnitLowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:282
[6] istril(A::Adjoint) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:285
[7] istril(A::Transpose) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:286
[8] istril(D::Diagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/diagonal.jl:131
[9] istril(M::Bidiagonal) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/bidiag.jl:250
[10] istril(A::SparseArrays.SparseMatrixCSC) in SparseArrays at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:3234
[11] istril(A::AbstractArray{T,2} where T) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162
[12] istril(A::AbstractArray{T,2} where T, k::Integer) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162
[13] istril(x::Number) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1171
[14] istril(::UniformScaling) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/uniformscaling.jl:104

In [28]:
@which istril(L)

istril(A::AbstractArray{T,2} where T) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/generic.jl:1162

---
If you open the file `julia/stdlib/v1.2/LinearAlgebra/src/generic.jl` and go to line 1162, you will see

```julia
function istril(A::AbstractMatrix, k::Integer = 0)
    require_one_based_indexing(A)
    m, n = size(A)
    for j in max(1, k + 2):n
        for i in 1:min(j - k - 1, m)
            iszero(A[i, j]) || return false
        end
    end
    return true
end
istril(x::Number) = true
```
---

In [29]:
#@edit istril(L)

In [30]:
n = 5
L = LowerTriangular(rand(n,n))

5×5 LowerTriangular{Float64,Array{Float64,2}}:
 0.964521    ⋅         ⋅         ⋅          ⋅      
 0.671654   0.996841   ⋅         ⋅          ⋅      
 0.517177   0.573759  0.591542   ⋅          ⋅      
 0.299998   0.383062  0.84302   0.0517429   ⋅      
 0.0950405  0.850467  0.542229  0.474453   0.360129

In [31]:
L.data

5×5 Array{Float64,2}:
 0.964521   0.0904625  0.0609714  0.598802   0.692204
 0.671654   0.996841   0.341189   0.416217   0.764619
 0.517177   0.573759   0.591542   0.62617    0.277068
 0.299998   0.383062   0.84302    0.0517429  0.122271
 0.0950405  0.850467   0.542229   0.474453   0.360129

In [32]:
L[1,2]

0.0

In [33]:
typeof(L)

LowerTriangular{Float64,Array{Float64,2}}

In [34]:
@which istril(L)

istril(A::LowerTriangular) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl:281

---
If you open the file `julia/stdlib/v1.2/LinearAlgebra/src/triangular.jl` and go to line 281, you will see

```julia
istril(A::LowerTriangular) = true
istril(A::UnitLowerTriangular) = true
istriu(A::UpperTriangular) = true
istriu(A::UnitUpperTriangular) = true
istril(A::Adjoint) = istriu(A.parent)
istril(A::Transpose) = istriu(A.parent)
istriu(A::Adjoint) = istril(A.parent)
istriu(A::Transpose) = istril(A.parent)
```
---

In [35]:
L'

5×5 Adjoint{Float64,LowerTriangular{Float64,Array{Float64,2}}}:
 0.964521  0.671654  0.517177  0.299998   0.0950405
 0.0       0.996841  0.573759  0.383062   0.850467 
 0.0       0.0       0.591542  0.84302    0.542229 
 0.0       0.0       0.0       0.0517429  0.474453 
 0.0       0.0       0.0       0.0        0.360129 

In [36]:
L = tril(rand(n,n))

5×5 Array{Float64,2}:
 0.838769    0.0        0.0       0.0       0.0    
 0.513315    0.876622   0.0       0.0       0.0    
 0.688326    0.0799426  0.547035  0.0       0.0    
 0.00574107  0.652881   0.217803  0.478076  0.0    
 0.015003    0.316821   0.25829   0.812704  0.45937

---

## Column-oriented forward substitution

Notice that the algorithm above is **row-oriented**, which is good for computer languages that store matrices in row-major order (like C/C++ or Python), but is bad for column-major order languages (like Fortran, MATLAB, or Julia).

We can produce a forward-substitution algorithm that is **column-oriented** by using the concept of **block-matrix multiplication**.

---

### Exercise:

Let the lower-triangular linear system $Lx = b$ be partitioned as

$$
\begin{bmatrix}
l_{11} & 0 \\
\hat{l} & \hat{L} \\
\end{bmatrix}
\begin{bmatrix}
x_1 \\ \hat{x}
\end{bmatrix} = 
\begin{bmatrix}
b_1 \\ \hat{b}
\end{bmatrix}
$$

where $\hat{L}$ is an $(n-1) \times (n-1)$ *lower-triangular* matrix and $\hat{x}$ and $\hat{b}$ are vectors of length $n - 1$.

1. Perform block-matrix multiplication.

2. Determine a _recursive algorithm_ for solving the linear system $Lx = b$.

3. Implement your recursive algorithm in `Julia` (name your function `col_rfs`).

4. Test your code for accuracy and efficiency.

---
### How to enter unicode characters:
- `\alpha + TAB = α`
- `l\hat + TAB = l̂`
- `x\_1 + TAB = x₁`
---

In [37]:
#Amy, Killian, Cameron
function col_rfs(L, b)
    n = size(L,1)
    
    l11 = L[1,1]
    b1 = b[1]
    lhat = L[2:n, 1]
    Lhat = L[2:n, 2:n]
    bhat = b[2:n]
    
    x1 = b1/l11
    bt = bhat - lhat*x1
    
    ##base case:
    if size(L,1) == 1
        return x1
    end
    
    #recursive case:
    xhat = col_rfs(Lhat, bt) 
    
    x = [x1; xhat]
    return x
    
end

L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

x = col_rfs(L, b)

4-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0

---
### Recursive algorithm

The partition

$$
\begin{bmatrix}
l_{11} & 0 \\
\hat{l} & \hat{L} \\
\end{bmatrix}
\begin{bmatrix}
x_1 \\ \hat{x}
\end{bmatrix} = 
\begin{bmatrix}
b_1 \\ \hat{b}
\end{bmatrix}
$$

gives us the following recursive algorithm:

1. $x_1 = b_1/l_{11}$
2. Solve $\hat{L} \hat{x} = \hat{b} - x_1 \hat{l}$ for $\hat{x}$.

---

### Exercise:

Use column-oriented forward substitution to solve the lower-triangular system.

$$
\begin{bmatrix}
2 & 0 & 0 & 0 \\
-1 & 2 & 0 & 0 \\
3 & 1 & -1 & 0 \\
4 & 1 & -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
2 \\ 3 \\ 2 \\ 9
\end{bmatrix}
$$


### Solution:

Then $x_1 = 2/2 = 1$ and

$$
\begin{bmatrix}
 2 & 0 & 0 \\
 1 & -1 & 0 \\
 1 & -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_2 \\ x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
3 \\ 2 \\ 9
\end{bmatrix}
- 
x_1 \cdot
\begin{bmatrix}
-1 \\ 3 \\ 4
\end{bmatrix}
=
\begin{bmatrix}
4 \\ -1 \\ 5
\end{bmatrix}.
$$

Then $x_2 = 4/2 = 2$ and

$$
\begin{bmatrix}
 -1 & 0 \\
 -3 & 3
\end{bmatrix}
\begin{bmatrix}
x_3 \\ x_4
\end{bmatrix} =
\begin{bmatrix}
-1 \\ 5
\end{bmatrix}
- 
x_2 \cdot
\begin{bmatrix}
1 \\ 1
\end{bmatrix}
=
\begin{bmatrix}
-3 \\ 3
\end{bmatrix}.
$$

Then $x_3 = -3/-1 = 3$ and

$$
\begin{bmatrix}
 3
\end{bmatrix}
\begin{bmatrix}
x_4
\end{bmatrix} =
\begin{bmatrix}
3
\end{bmatrix}
- 
x_3 \cdot
\begin{bmatrix}
-3
\end{bmatrix}
=
\begin{bmatrix}
12
\end{bmatrix}.
$$

Then $x_4 = 12/3 = 4$. Thus, $x = \begin{bmatrix} 1 \\ 2 \\ 3 \\ 4 \end{bmatrix}.$

---

### Exercise:

Inspired by your calculations in the previous exercise, write a non-recursive column-oriented forward-substitution code in `Julia` (name your function `col_fs`). Test your function for accuracy and efficiency.

```julia
# Caitlin, Killian, Sean
function col_fs(L, b)
    if !istril(L)
        error("L is not lower triangular")
    end
    n = size(L)[1]
    x = zeros(n)
    c = copy(b)
    for j in 1:n
        x[j] = c[j] / L[j,j]
        for i in j+1:n
            c[i] = c[i] - L[i,j]*x[j]
        end
    end
    return x
end
```

In [38]:
using LinearAlgebra

function row_fs(L, b)
    if !istril(L)
        error("L is not lower triangular")
    end
    n = size(L,1)
    x = copy(b)
    for i = 1:n
        for j = 1:i-1
            x[i] -= L[i,j]*x[j]
        end
        if L[i,i] == 0.0
            error("L is singular")
        end
        x[i] /= L[i,i]
    end
    return x
end

function col_fs(L, b)
    if !istril(L)
        error("L is not lower triangular")
    end
    n = size(L,1)
    x = copy(b)
    for j = 1:n
        if L[j,j] == 0.0
            error("L is singular")
        end
        x[j] /= L[j,j]
        for i = j+1:n
            x[i] -= L[i,j]*x[j]
        end
    end
    return x
end

L = [2 0 0 0; -1 2 0 0; 3 1 -1 0; 4 1 -3 3.0]
b = [2, 3, 2, 9.0]

@show x = row_fs(L, b)
@show x = col_fs(L, b);

x = row_fs(L, b) = [1.0, 2.0, 3.0, 4.0]
x = col_fs(L, b) = [1.0, 2.0, 3.0, 4.0]


---

### Exercise 3:

How do your `row_fs` and `col_fs` functions compare to `Julia`'s built-in `x = L\b`?

In [39]:
using BenchmarkTools

n = 1000
L = LowerTriangular(rand(n,n))
b = rand(n)

@btime x = row_fs(L,b)
@btime x = col_fs(L,b)
@btime x = L\b;

  1.099 ms (1 allocation: 7.94 KiB)
  714.426 μs (1 allocation: 7.94 KiB)
  616.124 μs (1 allocation: 7.94 KiB)


---